#include "questionlist.h"
#include "ui_questionlist.h"

#include "windows/newquestionwindow.h"
#include "model/question.h"
#include "model/answer.h"
#include "model/assignment.h"
#include "model/modality.h"
#include "DAO/daoassignmentsqlite.h"
#include "DAO/daoquestionsqlite.h"
#include "DAO/daoanswersqlite.h"
#include "DAO/daomodalitysqlite.h"
#include "DAO/daoanatomicalregionsqlite.h"

QuestionList::QuestionList(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::QuestionList)
{
    ui->setupUi(this);
    _btnDeleteItem = ui->deleteQuestionButton;
    _btnAddItem = ui->addNewQuestionButton;
    _view = ui->tableView;
    setWindowTitle("Quest√µes");

    QList<Modality> modalitiesList;
    DAOModality *daoModality = new DAOModalitySQLITE;
    modalitiesList = daoModality->getAllModalities();

    foreach (Modality item, modalitiesList) {
        ui->modalitiesComboBox->addItem(item.description(), item.id());
    }

    QPalette pal = _btnDeleteItem->palette();
    pal.setColor(QPalette::Button, QColor(Qt::red));
    _btnDeleteItem->setAutoFillBackground(true);
    _btnDeleteItem->setPalette(pal);
    _btnDeleteItem->update();
    QPalette pal2 = _btnAddItem->palette();
    pal2.setColor(QPalette::Button, QColor(Qt::green));
    _btnAddItem->setAutoFillBackground(true);
    _btnAddItem->setPalette(pal2);
    _btnAddItem->update();

    on_searchButton_clicked();
    delete daoModality;
}

QuestionList::~QuestionList()
{
    delete ui;
}

void QuestionList::editItem(QModelIndex model)
{
    QVariant description = model.data(0);
    QModelIndex sib = model.siblingAtColumn(1);
    QVariant id = sib.data(0).toInt();
    DAOQuestion *dao = new DAOQuestionSQLITE;
    Question question = dao->getQuestionById(id.toInt());
    NewQuestionWindow *newWindow = new NewQuestionWindow(&question);
    newWindow->show();
    this->close();
}

void QuestionList::on_modalitiesComboBox_currentIndexChanged(int index)
{
    int modalityId = ui->modalitiesComboBox->currentData().toInt();

    QList<AnatomicalRegion> anatomicalRegionList;
    DAOAnatomicalRegion *daoAnatomicalRegion = new DAOAnatomicalRegionSQLITE;
    anatomicalRegionList = daoAnatomicalRegion->getAnatomicalRegionByModalityId(modalityId);

    ui->anatomicalRegionComboBox->clear();
    foreach (AnatomicalRegion item, anatomicalRegionList) {
        ui->anatomicalRegionComboBox->addItem(item.description(), item.id());
    }

    delete daoAnatomicalRegion;
}

void QuestionList::on_anatomicalRegionComboBox_currentIndexChanged(int index)
{
    int anatomicalRegionId = ui->anatomicalRegionComboBox->currentData().toInt();

    QList<Assignment> assignmentList;
    DAOAssignment *assignmentDAO = new DAOAssignmentSQLITE;
    DAOQuestion *questionsDAO = new DAOQuestionSQLITE;
    assignmentList = assignmentDAO->getAssignmentsByAnatomicalRegion(anatomicalRegionId);

    ui->assignmentComboBox->clear();
    foreach (Assignment item, assignmentList) {
        item.setQuestionsList(questionsDAO->getQuestionsByAssignmentId(item.id()));
        ui->assignmentComboBox->addItem(item.description(), item.id());
    }
}

void QuestionList::on_addNewQuestionButton_clicked()
{
    NewQuestionWindow *newWindow = new NewQuestionWindow;
    newWindow->show();
    close();
}

void QuestionList::on_searchButton_clicked()
{
    int assignmentId = ui->assignmentComboBox->currentData().toInt();
    QList<Question> questionList;
    DAOQuestion *dao = new DAOQuestionSQLITE;
    questionList = dao->getQuestionsByAssignmentId(assignmentId);

    foreach (Question item, questionList) {
        _model.append(item);
    }

    _proxy.setSourceModel(&_model);
    _view->setModel(&_proxy);
    _view->resizeColumnsToContents();
    _view->setSelectionBehavior(QAbstractItemView::SelectRows);
    connect(_view, SIGNAL(doubleClicked(const QModelIndex &)), this, SLOT(editItem(QModelIndex)));

    delete dao;
}
